src/nextjournal/clerk/hashing.clj

```diff
@@ -199,40 +200,53 @@
 #_(parse-file {:markdown? true} "notebooks/rule_30.clj")
 #_(parse-file "notebooks/src/demo/lib.cljc")
 
+(defn- circular-dependency-error? [e]
+  (-> e ex-data :reason #{::dep/circular-dependency}))
+
+(defn- analyze-circular-dependency [state var form dep {:keys [node dependency]}]
+  (let [rec-form (concat '(do) [form (get-in state [:var->hash dependency :form])])
+        rec-var (symbol (str var "+" dep))]
+    (-> state
+        (update :graph #(-> %
+                            (dep/remove-edge dependency node)
+                            (dep/depend var rec-var)
+                            (dep/depend dep rec-var)))
+        (assoc-in [:var->hash rec-var :form] rec-form))))
+
+(defn- analyze-deps [var form {:as state :keys [graph]} dep]
+  (try (assoc state :graph (dep/depend graph (if var var form) dep))
+       (catch Exception e
+         (when-not (circular-dependency-error? e)
+           (throw e))
+         (analyze-circular-dependency state var form dep (ex-data e)))))
+
+(defn- analyze-codeblock [file state {:keys [type text]}]
+  (let [{:keys [var deps form ns-effect?]} (-> text read-string analyze)
+        state-with-var-hash (assoc-in state
+                                      [:var->hash (if var var form)]
+                                      {:file file
+                                       :form form
+                                       :deps deps})]
+    (when ns-effect?
+      (eval form))
```

Comment:
I don't fully understand the clerk workflow but in trying to use `clerk/show!` I found it was changing the ns of my repl. Would it be useful to capture and restore the current ns before running stuff ns-effectful expressions? Maybe similar to https://github.com/nextjournal/clerk/pull/26/files#diff-b33fbf28a0d44915162c46241d20d137f2ca22398ab87a4881637a3358159cc8R8-R12